### PR TITLE
Speech api

### PR DESCRIPTION
**Pull request by Josecisneros001**
This creates a new c++ node `AzureSpeechToText` that calls the [Azure API](https://azure.microsoft.com/es-mx/services/cognitive-services/speech-to-text/#get-started). For this, the hear.py in action_selectors check if there is internet connection, if yes, this node will be called and it published the result.

Information about using the sdk check its [github](https://github.com/Azure-Samples/cognitive-services-speech-sdk) and [this page](https://docs.microsoft.com/es-mx/azure/cognitive-services/speech-service/speech-sdk?tabs=linux).

### CODE REVIEW

File: catkin_home/src/action_selectors/scripts/hear.py [hunk 1]
```diff
@@ -26,6 +40,57 @@ def callback(data):
...
+    allsamples=[]
+    index=0
+    while True:
+        try:
+            index=index+1
+            sample=audioop.getsample(data,2,index)
+            sample_8a = sample & 0xff
+            sample_8b = (sample >> 8) & 0xff
+            allsamples.append(int(str(sample_8a)))
+            allsamples.append(int(str(sample_8b)))
```
Comment (diegocardozo97) [inline catkin_home/src/action_selectors/scripts/hear.py]: Never mind, I can't think of a better way to this.

File: .gitignore [hunk 2]
```diff
@@ -23,5 +23,6 @@
 *.DS_Store
 *.pyc
 .vscode/
-
+.vscode
```
Comment (diegocardozo97) [inline .gitignore]: Is this really necessary with `.vscode/`?

File: catkin_home/src/action_selectors/CMakeLists.txt [hunk 3]
```diff
@@ -154,6 +161,8 @@ include_directories(
 #   ${catkin_LIBRARIES}
 # )
 
+target_link_libraries(azureSpeechText_node  ${catkin_LIBRARIES} Microsoft.CognitiveServices.Speech.core )
```
[the remark this answers is not in the document]
Comment (Josecisneros001) [reply catkin_home/src/action_selectors/CMakeLists.txt]: It is the name of the .so file, it is inside "src/speechsdk/lib/x64"

File: catkin_home/src/action_selectors/scripts/hear.py [hunk 4]
```diff
@@ -26,6 +40,57 @@ def callback(data):
...
+    for i in data:
+        if last==-1:
+            last=i
+        else:
+            my_list_16b.append(struct.pack("B",int(last))+struct.pack("B",int(i)))
+            last=-1
+
+    return my_list_16b
+
+def getAllSamples(data):
```
Comment (SAlexisvf) [inline catkin_home/src/action_selectors/scripts/hear.py]: Be consistent with the naming format, `getAllSamples` --> `get_all_samples`

File: catkin_home/_GLOBAL_.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
+    "_comment":"This is a template of GLOBAL.json",
+    "AZURE_API_KEY":"",
+    "AZURE_REGION":""
+}
```
[the remark this answers is not in the document]
Comment (Josecisneros001) [reply catkin_home/_GLOBAL_.json]: Solved, I already replaced the json file with a txt file

File: catkin_home/src/action_selectors/src/.catkin_workspace [hunk 6]
```diff
@@ -0,0 +1 @@
+# This file currently only serves to mark the location of a catkin workspace for tool integration
```
[the remark this answers is not in the document]
Comment (Josecisneros001) [reply catkin_home/src/action_selectors/src/.catkin_workspace]: Solved, now it is excluded.

File: catkin_home/src/action_selectors/src/azureSpeechText.cpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        }
+    }
+}
+
+
+
+int main(int argc, char* argv[]){
+
+  // This must be called before anything else ROS-related
+  ros::init(argc, argv, "azureSpeechText_node");
```
[the remark this answers is not in the document]
Comment (Josecisneros001) [reply catkin_home/src/action_selectors/src/azureSpeechText.cpp]: Solved,now the name of the node is  "AzureSpeechToText"

File: .gitignore (hunk 2 above)
Comment (Josecisneros001) [reply .gitignore]: I didn't realize it was already there.

Comment (Josecisneros001) [reply .gitignore]: Solved, I already deleted one of the two files from the gitignore

File: catkin_home/src/action_selectors/scripts/hear.py [hunk 8]
```diff
@@ -9,8 +13,18 @@
...
 
 publisher = None
+publisher_16k = None
 asr_server = None
 
+def is_connected():
+    try:
+        # connect to the host -- tells us if the host is actually
+        # reachable
+        socket.create_connection(("www.google.com", 80))
```
[the remark this answers is not in the document]
Comment (Josecisneros001) [reply catkin_home/src/action_selectors/scripts/hear.py]: solved, i added a timeout to the connection and also a command to close it

File: catkin_home/src/action_selectors/scripts/hear.py [hunk 9]
```diff
@@ -26,6 +40,57 @@ def callback(data):
...
+def callback_azure(data):
+    rospy.loginfo("Received a voice audio, computing...")
+    
+    #Parse to int
+    dataOrd=[(ord(i)) for i in data.data]
+    
+    #GetSamples 2Bytes
+    Samples2B=getSamples2B(dataOrd)
+    #Change Sample Rate
+    resample=resample_ratecv(Samples2B,48000,16000)
```
[the remark this answers is not in the document]
Comment (Josecisneros001) [reply catkin_home/src/action_selectors/scripts/hear.py]: solved, I modified it to use bytearray

File: .gitignore [hunk 10]
```diff
@@ -9,13 +9,15 @@
...
 
-### DeepSpeech ###
+### DeepSpeech && Speech API ###
 /catkin_home/src/action_selectors/scripts/DeepSpeech/data/librispeech/manifest.dev-clean
 /catkin_home/src/action_selectors/scripts/DeepSpeech/data/librispeech/manifest.test-clean
 /catkin_home/src/action_selectors/scripts/DeepSpeech/models/baidu_en8k/mean_std.npz
 /catkin_home/src/action_selectors/scripts/DeepSpeech/models/baidu_en8k/params.tar.gz
 /catkin_home/src/action_selectors/scripts/DeepSpeech/models/baidu_en8k/vocab.txt
 /catkin_home/src/action_selectors/scripts/DeepSpeech/models/lm/lm.binary
+/catkin_home/src/action_selectors/src/.catkin_workspace
```
Comment (diegocardozo97) [inline .gitignore]: This line can be deleted.

File: .gitignore [hunk 11]
```diff
@@ -9,13 +9,15 @@
...
+GLOBAL.txt
```
Comment (diegocardozo97) [inline .gitignore]: Let's make this with the complete (absolute) path to the file.

**Pull request merged**